Guide to create an offline transaction (#315)
w3f/polkadot-wiki: docs/maintain-guides-how-to-create-offline-transaction.md

---
id: maintain-guides-how-to-create-offline-transaction
title: Generate an offline transaction
sidebar_label: Generate an offline transaction
---

# Generating an offline transaction

> There are few different ways to create an offline transaction by using the [@polkadot@cli-signer](https://github.com/polkadot-js/tools/tree/master/packages/signer-cli), [txwrapper](https://github.com/paritytech/txwrapper) or [Subkey](https://github.com/paritytech/substrate/tree/master/bin/utils/subkey). This guide shows you how to use each one of these for different scenarios. 

Reviewer: ```suggestion
There are few different ways to create an offline transaction: by using the [@polkadot@cli-signer](https://github.com/polkadot-js/tools/tree/master/packages/signer-cli), [txwrapper](https://github.com/paritytech/txwrapper), or [Subkey](https://github.com/paritytech/substrate/tree/master/bin/utils/subkey). This guide shows you how to use each one of these for different scenarios. 
```

## polkadot@cli-signer

> This would be the most easy to use if you are looking for CLI approach and have some basic understanding of the public-key cryptography. You can use that to create any transaction and submit it(Extrinsic) to the network directly or you create an offline transaction and then submit it over the RPC or [PolkadotJS UI](https://polkadot.js.org/apps/#/toolbox).

Reviewer: ```suggestion
This would be the easiest to use if you are looking for a CLI approach and have some basic understanding of public-key cryptography. You can use that to create any transaction and submit it (Extrinsic) to the network directly or you create an offline transaction and then submit it over RPC or [PolkadotJS UI](https://polkadot.js.org/apps/#/toolbox).
```

### Installation

> Assume you have installed both nodeJS and npm, you can run the following command:

Reviewer: ```suggestion
We assume you have installed [NodeJS with npm](https://nodejs.org). Run the following command:
```

> ```

Reviewer: Add language hint, bash

npm install -g @polkadot/api-cli

> ```

Reviewer: Add space after code fence

### Creating a transaction

Let's say we want to transfer 0.5 KSM from `DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX` to `EQ64DQpzBkMcPUqUh7ddvAWmuQFnX3gCPpY8E6F9EsZjHPT`. We need to construct a payload and then use `DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX` to sign it. 

Executing the following command:

> ```

Reviewer: Add language hint, bash

yarn run:signer sendOffline --account SENDER_ADDR --ws WEBSOCKET_ENDPOINT RUNTIME_MODULE.FUNCTION RECEIVER_ADDR AMOUNT
```

> ```

Reviewer: Add language hint, bash

yarn run:signer sendOffline --account DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX --ws wss://kusama-rpc.polkadot.io/ balances.transfer EQ64DQpzBkMcPUqUh7ddvAWmuQFnX3gCPpY8E6F9EsZjHPT 500000000000

```

You should see 

> ```

Reviewer: Add language hint, bash

Payload: 0x040050aab27d83bafddea5ffdb5d24fa38b98e3940d08c3744a65214cf7613930c51070088526a743501d0001a040000b0a8d493285c2df73290dfb7e61f870f17b41801197a149ca93654499ea3dafec37cfc20a7851d7139acd706cdc47a3abc6f8b23a28c7f72098138a09e15b4df
Signature> 
```

Then copy the payload and open another terminal to run the following command to sign it. 

> ```

Reviewer: Add language hint, bash

> Then copy the payload and open another terminal to run the following command to sign it. 

Reviewer 2: Just moving to another terminal does nothing.

Author: > This kind of guide, without appropriate context, makes me nervous.
> 
> Subkey is a no-go here.

oh..so Subkey does not allow to create an offline tx? I thought it is possible. 🤔

Just moving to another terminal does nothing. <--- Are you saying that it is not quite clear why you need to do that or the result is not what you expected?

Reviewer 2: > oh..so Subkey does not allow to create an offline tx? I thought it is possible.

Nope

> Are you saying that it is not quite clear why you need to do that or the result is not what you expected?

Read your instructions and think about it.

yarn run:signer sign --account SENDER_ADDR --seed SECRET_PHRASE --type KEYPAIR_CRYPTO_TYPE PAYLOAD
```

> ```

Reviewer: Add language hint, bash

yarn run:signer sign --account DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX --seed "normal salon one century nerve timber online engine actress walnut wheel display" --type sr25519 0x040050aab27d83bafddea5ffdb5d24fa38b98e3940d08c3744a65214cf7613930c51070088526a742503cc001a040000b0a8d493285c2df73290dfb7e61f870f17b41801197a149ca93654499ea3dafe6d6b1c0764c0e434e8f690052c72777c916d50aef8185cd5f40d6200c3a557cb

```

Result:

> ```

Reviewer: Add language hint, bash

Signature: 0x015ec182626a8407cc198d31efb14664e1f6d5b6cbe9190696e91df2c0c6ac1a2be7759be8533b56cd579a5f118dd3e490d7660abf40a3fafc9b10c4ae557ae584
```

Now you can copy & paste the signature to the first terminal. And you will have the signed transaction.

### Submitting to the network

```bash
Signed transaction:
0x390284226c361c136ed316babb2309f8e4fa4498939408868c3594f6257855808f666f015ec182626a8407cc198d31efb14664e1f6d5b6cbe9190696e91df2c0c6ac1a2be7759be8533b56cd579a5f118dd3e490d7660abf40a3fafc9b10c4ae557ae5843501d000040050aab27d83bafddea5ffdb5d24fa38b98e3940d08c3744a65214cf7613930c51070088526a74
```

> With that, you can submit the transaction via [PolkadotJS UI](https://polkadot.js.org/apps/#/toolbox).

Reviewer 2: Why are you submitting to Polkadot JS Apps? Clearly you already have an RPC connection to a node to start using this command. You could have just used `yarn run:signer submit` instead.

Author: yep, it is possible to use `yarn run:signer submit` to do this, but when I was working on that,  I wanted to let them know they could also use the Polkadotjs to submit the signed tx. will write down both options.

![offline transaction](assets/others/offline_transaction.jpg)

> One thing to keep in mind is that the transaction will be effective for five minutes. One of the potential use-cases would be a validator is sitting at the Intranet and the validator wants to keep changing the session keys for some periods of time and the only way to access the Internet is through the sentry node. Then the validator can submit the transaction via the RPC. 

Reviewer: ```suggestion
One thing to keep in mind is that the transaction will be effective for five minutes. One of the potential use-cases would be a validator is sitting at the Intranet and the validator wants to keep changing the session keys for some periods of time and the only way to access the Internet is through the sentry node. Then the validator can submit the transaction via RPC. 
```

> One thing to keep in mind is that the transaction will be effective for five minutes. One of the potential use-cases would be a validator is sitting at the Intranet and the validator wants to keep changing the session keys for some periods of time and the only way to access the Internet is through the sentry node. Then the validator can submit the transaction via the RPC. 

Reviewer 2: ?

---
id: maintain-guides-how-to-create-offline-transaction
title: Generate an offline transaction
sidebar_label: Generate an offline transaction
---

# Generating an offline transaction

There are few different ways to create an offline transaction: by using the [@polkadot@cli-signer](https://github.com/polkadot-js/tools/tree/master/packages/signer-cli), [txwrapper](https://github.com/paritytech/txwrapper), or [Subkey](https://github.com/paritytech/substrate/tree/master/bin/utils/subkey). This guide shows you how to use each one of these for different scenarios. 

## polkadot@cli-signer

> This would be the easiest to use if you are looking for a CLI approach and have some basic understanding of public-key cryptography. You can use that to create any transaction and submit it (Extrinsic) to the network directly or you create an offline transaction and then submit it over RPC or [PolkadotJS UI](https://polkadot.js.org/apps/#/toolbox).

Reviewer 3: ```suggestion
This would be the easiest to use if you are looking for a CLI approach and have some basic understanding of public-key cryptography. You can use that to create an extrinsic and submit it as a transaction to the network directly or you sign an offline transaction and submit it over RPC or on [PolkadotJS Apps UI](https://polkadot.js.org/apps/#/toolbox).
```

> 

Reviewer 3: ```suggestion
```

double space

### Installation

We assume you have installed [NodeJS with npm](https://nodejs.org). Run the following command:

```bash
npm install -g @polkadot/api-cli
```

### Creating a transaction

Let's say we want to transfer 0.5 KSM from `DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX` to `EQ64DQpzBkMcPUqUh7ddvAWmuQFnX3gCPpY8E6F9EsZjHPT`. We need to construct a payload and then use `DMTHrNcmA8QbqRS4rBq8LXn8ipyczFoNMb1X4cY2WD9tdBX` to sign it. 

Executing the following command:

```bash

> yarn run:signer sendOffline --account SENDER_ADDR --ws WEBSOCKET_ENDPOINT RUNTIME_MODULE.FUNCTION RECEIVER_ADDR AMOUNT

Reviewer 3: Doesn't using `yarn run:signer` imply you've cloned the repository locally? In the previous step you installed it from npm, shouldn't you be using the command provided in the path?

Reviewer: Yes please default to yarn for everything - it works better on Windows and in general.

Author: good catch!
